enterprise/server/scheduling/scheduler_server/scheduler_server.go

```diff
@@ -1662,95 +1678,147 @@ func (s *SchedulerServer) enqueueTaskReservations(ctx context.Context, enqueueRe
 		}
-		if sampleIndex == 0 {
-			if preferredNode != nil {
-				nodes = []*executionNode{preferredNode}
-				// Unset the preferred node so that we fall back to random sampling
-				// (in subsequent loop iterations) if the preferred node probe fails.
-				preferredNode = nil
-			} else {
-				nodes = nodeBalancer.GetNodes(opts.scheduleOnConnectedExecutors)
-				if len(nodes) == 0 {
-					return status.UnavailableErrorf("No registered executors in pool %q with os %q with arch %q.", pool, os, arch)
-				}
-				nodes = nodesThatFit(nodes, enqueueRequest.GetTaskSize())
-				if len(nodes) == 0 {
-					return status.UnavailableErrorf(
-						"No registered executors in pool %q with os %q with arch %q can fit a task with %d milli-cpu and %d bytes of memory.",
-						pool, os, arch,
-						enqueueRequest.GetTaskSize().GetEstimatedMilliCpu(),
-						enqueueRequest.GetTaskSize().GetEstimatedMemoryBytes())
-				}
-				rankedNodes := s.taskRouter.RankNodes(ctx, cmd, remoteInstanceName, toNodeInterfaces(nodes))
-				nodes, err = fromNodeInterfaces(rankedNodes)
-				if err != nil {
-					return err
-				}
-			}
-		}
-		if sampleIndex >= len(nodes) {
-			return status.FailedPreconditionErrorf("sampleIndex %d >= %d", sampleIndex, len(nodes))
-		}
-		node := nodes[sampleIndex]
-		sampleIndex = (sampleIndex + 1) % len(nodes)
+
 		// Set the executor ID in case the node is owned by another scheduler, so
 		// that the scheduler can prefer this node for the probe.
-		enqueueRequest.ExecutorId = node.GetExecutorID()
+		enqueueRequest.ExecutorId = rankedNode.node.GetExecutorID()
 
 		enqueueStart := time.Now()
-		if opts.scheduleOnConnectedExecutors {
-			if node.handle == nil {
-				log.CtxErrorf(ctx, "nil handle for a local executor %q", node.GetExecutorID())
-				continue
-			}
-			_, err := node.handle.EnqueueTaskReservation(ctx, enqueueRequest)
-			if err != nil {
-				continue
-			}
+		if scheduledOnPreferredNode && !rankedNode.preferred && nonPreferredDelay > 0*time.Second {
+			enqueueRequest.Delay = durationpb.New(nonPreferredDelay)
 		} else {
-			if node.schedulerHostPort == "" {
-				log.CtxErrorf(ctx, "node %q has no scheduler host:port set", node.GetExecutorID())
-				continue
-			}
-
-			schedulerClient, err := s.schedulerClientCache.get(node.schedulerHostPort)
-			if err != nil {
-				log.CtxWarningf(ctx, "Could not get SchedulerClient for %q: %s", node.schedulerHostPort, err)
-				continue
-			}
-			rpcCtx, cancel := context.WithTimeout(ctx, schedulerEnqueueTaskReservationTimeout)
-			_, err = schedulerClient.EnqueueTaskReservation(rpcCtx, enqueueRequest)
-			cancel()
-			if err != nil {
-				log.CtxWarningf(ctx, "EnqueueTaskReservation via scheduler target %q failed: %s", node.schedulerHostPort, err)
-				time.Sleep(schedulerEnqueueTaskReservationFailureSleep)
-				continue
+			enqueueRequest.Delay = nil
+		}
+		enqueued, err := s.enqueue(ctx, rankedNode.node, enqueueRequest, opts)
+		if err != nil {
+			return err
```

Comment:
I don't think you meant to return an error here. This would mean we stop attempting enqueing after the first error. 

Now that I read the code a little more I see that enqueue only returns an error if the context is done. Still this is very error prone because if someone ever modifies enqueue to return an error in other conditions, it will cause us to stop enqueing early. If you have any termination conditions you want to check, I would move them to this loop.

Reply:
Good catch, this is indeed bug-prone. Removed the error return from the called function and added a test that should catch the case you mentioned.